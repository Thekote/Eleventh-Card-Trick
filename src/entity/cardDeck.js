const SUITS = ["Hearts", "Diamonds", "Spades", "Clubs"];
const VALUES = [
  "A",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "J",
  "Q",
  "K",
];

function createCardDeck() {
  return SUITS.flatMap((suit) => VALUES.map((value) => ({ suit, value })));
}
const cardDeck = createCardDeck();

function shuffleDeck(deck) {
  const currentDeck = [...deck];
  const shuffledDeck = currentDeck.sort(() => 0.5 - Math.random());

  return shuffledDeck;
}

export { createCardDeck, cardDeck, shuffleDeck };
