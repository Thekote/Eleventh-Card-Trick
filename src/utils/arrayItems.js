function getArrayItems(items) {
  return items.slice(0, 21);
}

export { getArrayItems };
