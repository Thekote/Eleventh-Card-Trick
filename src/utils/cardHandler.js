const SUITS = ["Hearts", "Diamonds", "Spades", "Clubs"];
const VALUES = [
  "A",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "J",
  "Q",
  "K",
];

function createCardDeck() {
  return SUITS.flatMap((suit) => {
    return VALUES.map((value) => {
      return { suit, value };
    });
  });
}
const cardDeck = createCardDeck();

function shuffleDeck(deck) {
  const currentDeck = [...deck];
  const shuffledDeck = currentDeck.sort(() => 0.5 - Math.random());

  return shuffledDeck;
}

function getFirst21Items(items) {
  return items.slice(0, 21);
}

export { cardDeck, shuffleDeck, getFirst21Items };
