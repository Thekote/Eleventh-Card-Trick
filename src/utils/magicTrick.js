function magicTrick(array, rowIndex) {
  const arrayCopy = [...array];
  const magicArray = [
    arrayCopy[rowIndex === 0 ? 1 : 0],
    arrayCopy[rowIndex],
    arrayCopy[rowIndex === 2 ? 1 : 2],
  ].flat();

  return magicArray.reduce(
    (acc, element, index) => {
      if (index % 3 === 0) {
        return [[element, ...acc[0]], acc[1], acc[2]];
      } else if (index % 3 === 1) {
        return [acc[0], [element, ...acc[1]], acc[2]];
      } else if (index % 3 === 2) {
        return [acc[0], acc[1], [element, ...acc[2]]];
      }
    },
    [[], [], []]
  );
}

export { magicTrick };
