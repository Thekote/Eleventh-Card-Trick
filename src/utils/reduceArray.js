function reduceArray(array, elementIndex) {
  const arrayCopy = [...array];
  const makeReducedArray = [
    arrayCopy[elementIndex === 0 ? 1 : 0],
    arrayCopy[elementIndex],
    arrayCopy[elementIndex === 2 ? 1 : 2],
  ].flat();
  return makeReducedArray.reduce(
    // eslint-disable-next-line array-callback-return
    (acc, element, index) => {
      if (index % 3 === 0) {
        return [[element, ...acc[0]], acc[1], acc[2]];
      } else if (index % 3 === 1) {
        return [acc[0], [element, ...acc[1]], acc[2]];
      } else if (index % 3 === 2) {
        return [acc[0], acc[1], [element, ...acc[2]]];
      }
    },
    [[], [], []]
  );
}

export { reduceArray };
