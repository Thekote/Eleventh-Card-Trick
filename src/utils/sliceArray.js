function sliceArray(array) {
  const fistArrayElement = array.slice(0, 7);
  const secondArrayElement = array.slice(7, 14);
  const thirdArrayElement = array.slice(14, 22);

  return [fistArrayElement, secondArrayElement, thirdArrayElement];
}

export { sliceArray };
