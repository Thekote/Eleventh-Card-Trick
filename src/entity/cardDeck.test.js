import { cardDeck, shuffleDeck } from "./cardDeck";
import { getArrayItems } from "../utils/arrayItems";

describe("cardDeck", () => {
  it("should return the created deck", () => {
    expect(cardDeck).toMatchSnapshot();
  });
});

describe("shuffleDeck", () => {
  it("should return a shuffled deck", () => {
    expect(cardDeck).not.toStrictEqual(shuffleDeck(cardDeck));
  });
});

describe("getArrayItems", () => {
  it("should return the first 21 items", () => {
    const initalItems = [
      1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
      22, 23, 24, 25,
    ];
    const expectedItems = [
      1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
    ];

    expect(getArrayItems(initalItems)).toEqual(expectedItems);
  });
});
