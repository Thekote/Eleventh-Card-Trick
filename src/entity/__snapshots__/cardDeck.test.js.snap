// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cardDeck should return the created deck 1`] = `
Array [
  Object {
    "suit": "Hearts",
    "value": "A",
  },
  Object {
    "suit": "Hearts",
    "value": "2",
  },
  Object {
    "suit": "Hearts",
    "value": "3",
  },
  Object {
    "suit": "Hearts",
    "value": "4",
  },
  Object {
    "suit": "Hearts",
    "value": "5",
  },
  Object {
    "suit": "Hearts",
    "value": "6",
  },
  Object {
    "suit": "Hearts",
    "value": "7",
  },
  Object {
    "suit": "Hearts",
    "value": "8",
  },
  Object {
    "suit": "Hearts",
    "value": "9",
  },
  Object {
    "suit": "Hearts",
    "value": "10",
  },
  Object {
    "suit": "Hearts",
    "value": "J",
  },
  Object {
    "suit": "Hearts",
    "value": "Q",
  },
  Object {
    "suit": "Hearts",
    "value": "K",
  },
  Object {
    "suit": "Diamonds",
    "value": "A",
  },
  Object {
    "suit": "Diamonds",
    "value": "2",
  },
  Object {
    "suit": "Diamonds",
    "value": "3",
  },
  Object {
    "suit": "Diamonds",
    "value": "4",
  },
  Object {
    "suit": "Diamonds",
    "value": "5",
  },
  Object {
    "suit": "Diamonds",
    "value": "6",
  },
  Object {
    "suit": "Diamonds",
    "value": "7",
  },
  Object {
    "suit": "Diamonds",
    "value": "8",
  },
  Object {
    "suit": "Diamonds",
    "value": "9",
  },
  Object {
    "suit": "Diamonds",
    "value": "10",
  },
  Object {
    "suit": "Diamonds",
    "value": "J",
  },
  Object {
    "suit": "Diamonds",
    "value": "Q",
  },
  Object {
    "suit": "Diamonds",
    "value": "K",
  },
  Object {
    "suit": "Spades",
    "value": "A",
  },
  Object {
    "suit": "Spades",
    "value": "2",
  },
  Object {
    "suit": "Spades",
    "value": "3",
  },
  Object {
    "suit": "Spades",
    "value": "4",
  },
  Object {
    "suit": "Spades",
    "value": "5",
  },
  Object {
    "suit": "Spades",
    "value": "6",
  },
  Object {
    "suit": "Spades",
    "value": "7",
  },
  Object {
    "suit": "Spades",
    "value": "8",
  },
  Object {
    "suit": "Spades",
    "value": "9",
  },
  Object {
    "suit": "Spades",
    "value": "10",
  },
  Object {
    "suit": "Spades",
    "value": "J",
  },
  Object {
    "suit": "Spades",
    "value": "Q",
  },
  Object {
    "suit": "Spades",
    "value": "K",
  },
  Object {
    "suit": "Clubs",
    "value": "A",
  },
  Object {
    "suit": "Clubs",
    "value": "2",
  },
  Object {
    "suit": "Clubs",
    "value": "3",
  },
  Object {
    "suit": "Clubs",
    "value": "4",
  },
  Object {
    "suit": "Clubs",
    "value": "5",
  },
  Object {
    "suit": "Clubs",
    "value": "6",
  },
  Object {
    "suit": "Clubs",
    "value": "7",
  },
  Object {
    "suit": "Clubs",
    "value": "8",
  },
  Object {
    "suit": "Clubs",
    "value": "9",
  },
  Object {
    "suit": "Clubs",
    "value": "10",
  },
  Object {
    "suit": "Clubs",
    "value": "J",
  },
  Object {
    "suit": "Clubs",
    "value": "Q",
  },
  Object {
    "suit": "Clubs",
    "value": "K",
  },
]
`;
